File: sdk-java/src/main/java/io/daytona/sdk/VolumeService.java
// Copyright Daytona Platforms Inc.
// SPDX-License-Identifier: Apache-2.0

package io.daytona.sdk;

import io.daytona.api.client.api.VolumesApi;
import io.daytona.api.client.model.CreateVolume;
import io.daytona.api.client.model.HotmountRegion;
import io.daytona.api.client.model.Region;
import io.daytona.api.client.model.VolumeMountTokenDto;
import io.daytona.api.client.model.VolumeType;
import io.daytona.sdk.model.Volume;

import java.util.List;
import java.util.ArrayList;

/**
 * Service for managing Daytona Volumes.
 *
 * <p>Volumes provide persistent shared storage that can be mounted into Sandboxes.
 */
public class VolumeService {
    private final VolumesApi volumesApi;

    VolumeService(VolumesApi volumesApi) {
        this.volumesApi = volumesApi;
    }

    /**
     * Creates a new legacy volume.
     *
     * @param name volume name
     * @return created {@link Volume}
     * @throws io.daytona.sdk.exception.DaytonaException if creation fails
     */
    public Volume create(String name) {
        return create(name, null, null);
    }

    /**
     * Creates a new volume of the given type.
     *
     * @param name volume name
     * @param type volume type, or null to default to legacy
     * @return created {@link Volume}
     * @throws io.daytona.sdk.exception.DaytonaException if creation fails
     */
    public Volume create(String name, VolumeType type) {
        return create(name, type, (String) null);
    }

    /**
     * Creates a new volume of the given type, with a region.
     *
     * @param name volume name
     * @param type volume type, or null to default to legacy
     * @param region region to create the volume in; for blockmount volumes it selects the region-local
     *     store the volume's data lives in — a performance/placement knob, sandboxes in any region can
     *     attach it (optional for blockmount, defaults to the organization's default region when omitted),
     *     selects the deployment region for hotmount volumes, not allowed for legacy
     * @return created {@link Volume}
     * @throws io.daytona.sdk.exception.DaytonaException if creation fails
     */
    public Volume create(String name, VolumeType type, String region) {
        io.daytona.api.client.model.VolumeDto volumeDto = ExceptionMapper.callMain(
                () -> volumesApi.createVolume(
                        new CreateVolume().name(name).type(type).region(region), null)
        );
        return toVolume(volumeDto);
    }

    /**
     * Creates a short-lived mount token for a hotmount volume.
     *
     * <p>The token, together with the returned region gateway/binaries endpoints, is used to
     * bootstrap the hotmount agent (inside a Sandbox or on customer infrastructure) and mount the
     * volume on the fly. Only hotmount volumes support this.
     *
     * @param volumeId the hotmount volume identifier
     * @return the mount token, region endpoints, and expiration
     * @throws io.daytona.sdk.exception.DaytonaException if the request fails
     */
    public VolumeMountTokenDto getMountToken(String volumeId) {
        return ExceptionMapper.callMain(() -> volumesApi.createVolumeMountToken(volumeId, null, null));
    }

    /**
     * Lists the hotmount regions available for volume creation.
     *
     * @return list of active hotmount regions
     * @throws io.daytona.sdk.exception.DaytonaException if the request fails
     */
    public List<HotmountRegion> listHotmountRegions() {
        List<HotmountRegion> regions = ExceptionMapper.callMain(() -> volumesApi.listHotmountRegions(null));
        return regions != null ? regions : new ArrayList<HotmountRegion>();
    }

    /**
     * Lists the regions where blockmount volumes can be created.
     *
     * <p>A blockmount volume's data lives in the region it is created in (a performance/placement knob —
     * sandboxes in any region can attach it, colocation is just faster). Only regions a superadmin has
     * enabled for blockmount are returned.
     *
     * @return list of regions that support blockmount volumes
     * @throws io.daytona.sdk.exception.DaytonaException if the request fails
     */
    public List<Region> listBlockmountRegions() {
        List<Region> regions = ExceptionMapper.callMain(() -> volumesApi.listBlockmountRegions(null));
        return regions != null ? regions : new ArrayList<Region>();
    }

    /**
     * Lists all accessible volumes.
     *
     * @return list of available volumes
     * @throws io.daytona.sdk.exception.DaytonaException if the API request fails
     */
    public List<Volume> list() {
        List<io.daytona.api.client.model.VolumeDto> volumes = ExceptionMapper.callMain(() -> volumesApi.listVolumes(null, null));
        List<Volume> result = new ArrayList<Volume>();
        if (volumes != null) {
            for (io.daytona.api.client.model.VolumeDto volume : volumes) {
                result.add(toVolume(volume));
            }
        }
        return result;
    }

    /**
     * Retrieves a volume by name.
     *
     * @param name volume name
     * @return matching {@link Volume}
     * @throws io.daytona.sdk.exception.DaytonaException if no volume is found or request fails
     */
    public Volume getByName(String name) {
        io.daytona.api.client.model.VolumeDto volumeDto = ExceptionMapper.callMain(() -> volumesApi.getVolumeByName(name, null));
        return toVolume(volumeDto);
    }

    /**
     * Deletes a volume by ID.
     *
     * @param id volume identifier
     * @throws io.daytona.sdk.exception.DaytonaException if deletion fails
     */
    public void delete(String id) {
        ExceptionMapper.runMain(() -> volumesApi.deleteVolume(id, null));
    }

    private Volume toVolume(io.daytona.api.client.model.VolumeDto source) {
        Volume volume = new Volume();
        if (source != null) {
            volume.setId(source.getId());
            volume.setName(source.getName());
            volume.setType(source.getType() == null ? null : source.getType().getValue());
            volume.setSizeInGb(source.getSizeInGb());
            volume.setRegion(source.getRegion());
            volume.setShared(source.getShared());
            volume.setState(source.getState() == null ? null : source.getState().getValue());
        }
        return volume;
    }
}

File: sdk-ruby/lib/daytona/volume_service.rb
# Copyright Daytona Platforms Inc.
# SPDX-License-Identifier: Apache-2.0

# frozen_string_literal: true

module Daytona
  class VolumeService
    include Instrumentation

    # Service for managing Daytona Volumes. Can be used to list, get, create and delete Volumes.
    #
    # @param volumes_api [DaytonaApiClient::VolumesApi]
    # @param otel_state [Daytona::OtelState, nil]
    def initialize(volumes_api, otel_state: nil)
      @volumes_api = volumes_api
      @otel_state = otel_state
    end

    # Create new Volume.
    #
    # @param name [String]
    # @param type [String, nil] The volume type. Defaults to legacy.
    # @param region [String, nil] The region to create the Volume in. For blockmount volumes it selects
    #   the region-local store the Volume's data lives in — a performance/placement knob; Sandboxes in any
    #   region can attach it. Optional for blockmount, defaulting to the organization's default region when
    #   omitted. Selects the deployment region for hotmount volumes.
    # @return [Daytona::Volume]
    def create(name, type: nil, region: nil)
      Volume.new(volumes_api.create_volume(DaytonaApiClient::CreateVolume.new(name:, type:, region:)))
    end

    # Delete a Volume.
    #
    # @param volume [Daytona::Volume]
    # @return [void]
    def delete(volume) = volumes_api.delete_volume(volume.id)

    # Get a Volume by name.
    #
    # @param name [String]
    # @param create [Boolean]
    # @param type [String, nil] The volume type to create (only used if create is true).
    # @param region [String, nil] The region to create the Volume in. For blockmount volumes it selects
    #   the region-local store the Volume's data lives in — a performance/placement knob; Sandboxes in any
    #   region can attach it. Optional for blockmount, defaulting to the organization's default region when
    #   omitted. Selects the deployment region for hotmount volumes.
    # @return [Daytona::Volume]
    def get(name, create: false, type: nil, region: nil)
      Volume.new(volumes_api.get_volume_by_name(name))
    rescue DaytonaApiClient::ApiError => e
      raise unless create && e.code == 404 && e.message.include?("Volume with name #{name} not found")

      create(name, type:, region:)
    end

    # Create a short-lived mount token for a hotmount Volume.
    #
    # The token, together with the returned region gateway/binaries endpoints, is used to
    # bootstrap the hotmount agent (inside a Sandbox or on customer infrastructure) and mount
    # the Volume on the fly. Only hotmount volumes support this.
    #
    # @param volume [Daytona::Volume]
    # @return [DaytonaApiClient::VolumeMountTokenDto]
    def get_mount_token(volume) = volumes_api.create_volume_mount_token(volume.id)

    # List the hotmount regions available for volume creation.
    #
    # @return [Array<DaytonaApiClient::HotmountRegion>]
    def list_hotmount_regions = volumes_api.list_hotmount_regions

    # List the regions where blockmount Volumes can be created. A blockmount Volume's data lives in
    # the region it is created in (a performance/placement knob — Sandboxes in any region can attach
    # it, colocation is just faster).
    #
    # @return [Array<DaytonaApiClient::Region>]
    def list_blockmount_regions = volumes_api.list_blockmount_regions

    # List all Volumes.
    #
    # @return [Array<Daytona::Volume>]
    def list
      volumes_api.list_volumes.map { |volume| Volume.new(volume) }
    end

    instrument :create, :delete, :get, :get_mount_token, :list_hotmount_regions, :list_blockmount_regions, :list,
               component: 'VolumeService'

    private

    # @return [DaytonaApiClient::VolumesApi]
    attr_reader :volumes_api

    # @return [Daytona::OtelState, nil]
    attr_reader :otel_state
  end
end
